esrally/telemetry.py

```diff
@@ -907,6 +910,13 @@ def jvm_mem_stats(self, node_name, node_stats):
     def os_mem_stats(self, node_name, node_stats):
         return self.flatten_stats_fields(prefix="os_mem", stats=node_stats["os"]["mem"])
 
+    def os_cgroup_stats(self, node_name, node_stats):
+        # Convert strings returned by the Node Stats API for os.cgroup.memory limits
+        # https://github.com/elastic/elasticsearch/issues/93429
+        for k in ("limit_in_bytes", "usage_in_bytes"):
+            node_stats["os"]["cgroup"]["memory"].update({k: int(node_stats["os"]["cgroup"]["memory"].get(k))})
```

Comment:
This doesn't look safe to me. First of all I see that `node-stats-include-cgroup` by default is set to `true`, so it'll try to collect those stats always when the telemetry device is enabled.

But then you are assuming that the dict `node_stats` already has the keys `os` / `group` / `memory` here.

In the logs against a fresh `8.6.1` cluster I saw the following warning:

```
2023-02-06 11:02:25,752 ActorAddr-(T|:37073)/PID:439715 esrally.telemetry ERROR Could not determine node stats
Traceback (most recent call last):

  File "/home/dl/source/elastic/rally/esrally/telemetry.py", line 172, in run
    self.recorder.record()

  File "/home/dl/source/elastic/rally/esrally/telemetry.py", line 850, in record
    collected_node_stats.update(self.os_cgroup_stats(node_name, node_stats))

  File "/home/dl/source/elastic/rally/esrally/telemetry.py", line 917, in os_cgroup_stats
    node_stats["os"]["cgroup"]["memory"].update({k: int(node_stats["os"]["cgroup"]["memory"].get(k))})

ValueError: invalid literal for int() with base 10: 'max'

```

When I did some test runs against a real metric store I didn't find any os cgroup related stats (target ES running in docker 8.6.1).

When we aren't sure that a dict has keys present, we should use the [setdefault()](https://docs.python.org/3/library/stdtypes.html#dict.setdefault) method.

Reply:
Would you mind to share the output of `GET /_nodes/stats?filter_path=**.os.cgroup` from your 8.6.1 cluster? I did not encounter any values named `max`, though it looks like ES will try [fetching it](https://github.com/elastic/elasticsearch/blob/5649ec2705afe7ef6e20937761ef50c857e755d9/server/src/main/java/org/elasticsearch/monitor/os/OsProbe.java#L540) in the form of `memory.max` from cgroup fs along with `memory.current`.

I'll need to rework this and fix the tests.

Reply:
> Would you mind to share the output of `GET /_nodes/stats?filter_path=**.os.cgroup` from your 8.6.1 cluster? I did not encounter any values named `max`, though it looks like ES will try [fetching it](https://github.com/elastic/elasticsearch/blob/5649ec2705afe7ef6e20937761ef50c857e755d9/server/src/main/java/org/elasticsearch/monitor/os/OsProbe.java#L540) in the form of `memory.max` from cgroup fs along with `memory.current`.

sure thing. Here you go:

```
{
  "nodes" : {
    "yPfEvz0MSHS1A9Oq4YBWzQ" : {
      "os" : {
        "cgroup" : {
          "cpuacct" : {
            "control_group" : "/",
            "usage_nanos" : 62952671
          },
          "cpu" : {
            "control_group" : "/",
            "cfs_period_micros" : 100000,
            "cfs_quota_micros" : -1,
            "stat" : {
              "number_of_elapsed_periods" : 0,
              "number_of_times_throttled" : 0,
              "time_throttled_nanos" : 0
            }
          },
          "memory" : {
            "control_group" : "/",
            "limit_in_bytes" : "max",
            "usage_in_bytes" : "35197526016"
          }
        }
      }
    }
  }
}
```